### DISCUSSION TITLE
consecutive fireDocumentUpdate

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1562] [Q&A]
hi, what will happen if a second fireDocumentUpdate is emitting while the first is still running.
will the 2nd one wait? the first one cancelled?

**Comment by msujew** [comment]
Hey @cdietrich,

if you call it in the context of a `WorkspaceLock#write` operation, the first one will be cancelled. If you don't use the `WorkspaceLock`, then it depends on whether you call it inside of a `DocumentBuilder#onBuildPhase` call. If so, the inner will complete first, then the outer one. If you don't then both will run in parallel.

I would recommend you to call the `WorkspaceLock#write` method though.

**Reply by cdietrich** [reply]
@msujew we dont call anything.
we have the following problem:
we have a vscode (codeserver) in the browser with a lazy file system.
even after this one has setteled we see reference errors in langium based files that are not real errors
so i wonder what happens:
- the sync provides the first 100 files, fireDocumentUpdate is called by didChangeWatchedFiles
- while these files are still processed, the next 100 files are reported by the watcher.
will we now stop processing the last 25 files of the first 100 that were not processed yet.

**Reply by cdietrich** [reply]
of course this problem happens very rearely

**Reply by msujew** [reply]
Ah, I see. However, I'm a bit confused by the behavior. We have written this specifically with this case in mind. Basically, whenever you call `DocumentBuilder#update` (which happens on every document update from the language client), all new documents are registered before the [first opportunity to cancel](https://github.com/eclipse-langium/langium/blob/d83d32e75c820cf20ea1e2b59d33cf3145cf7264/packages/langium/src/workspace/document-builder.ts#L210) arrives. The `DocumentBuilder` is supposed to be written in such a way that no document is ever lost when we cancel the build process.

So yes, while the build process is interrupted when the next 100 files are reported by the watcher, the second update call should then process all files that have not finished the document lifecycle yet.

**Reply by cdietrich** [reply]
I guess we need to add more logging or instrumenting and debugging to understand what is really happening

**Reply by cdietrich** [reply]
Maybe also a problem with affected calculation and new files

**Reply by msujew** [reply]
I'd appreciate a repro example/test in case you manage to reproduce the behavior with plain Langium.

**Reply by cdietrich** [reply]
i am at least getting into the realm of reproducing.
problem also happens independent of cancelling

**Reply by cdietrich** [reply]
hmmm we see a

```
protected shouldRelink(document: LangiumDocument, changedUris: Set<string>): boolean {
        // Relink documents with linking errors -- maybe those references can be resolved now
        if (document.references.some(ref => ref.error !== undefined)) {
            return true;
        }
```
        returning false.
        
        should that not ever ever return true as long as there are errors on the document?

**Reply by cdietrich** [reply]
what i did not check if the file is actually ok but the marker removal might be broken

**Reply by cdietrich** [reply]
added some more logging. a document that was in build phase linked did not end up in build phase validated.
thus no marker removal

**Reply by cdietrich** [reply]
@msujew what code is intended to reset the document state in case it needs to be reevaluated cause affected.
doc.state = Math.min(doc.state, DocumentState.ComputedScopes);
correct?

**Reply by msujew** [reply]
Yes, exactly. See for more info:

https://github.com/eclipse-langium/langium/blob/d83d32e75c820cf20ea1e2b59d33cf3145cf7264/packages/langium/src/workspace/document-builder.ts#L197-L206

**Reply by cdietrich** [reply]
hmmmmm i still have a hard time to figure out what is going on. 
i can reproduce it more or less in 1-4 cases.
but i am not sure if i have any good idea or i am on the correct path.
do you have access to a dsl with sample projects with more than a few files. if yes i could provide you my "toucher"

**Reply by cdietrich** [reply]
what i can see is this

```
validate file:///Users/dietrich/Downloads/all_end_june/xxxx/yyyy.zzzz
fireDocumentUpdate changed: 5 deleted: 0
update: 5 deleted: 0
emitUpdate changed: 5 deleted: 0
onBuildPhase Parsed [
```
so during validation the build seems to be cancelled.
now on the next validation

shouldValidate file:///Users/dietrich/Downloads/all_end_june/xxxx/yyyy.zzzz => true

but

skipping file:///Users/dietrich/Downloads/all_end_june/xxxx/yyyy.zzzz cause already in validated state

=> how is the notification to end up in the client

await this.notifyBuildPhase(filtered, targetState, cancelToken)
will ignore it

(maybe also an add. problem with caches not be evicted)

**Reply by cdietrich** [reply]
do we need a notified state?
or emit the validated event for
the filtered (!= validated) and the ones that are not completed yet?

**Reply by cdietrich** [reply]
other states might have the same problem.we advance the state. but in doubt would never notify

**Reply by cdietrich** [reply]
possible reproducer:
https://github.com/eclipse-langium/langium/compare/main...cdietrich:langium:cd-reproducer

**Reply by cdietrich** [reply]
another possible solution might be to undo state processes in case a cancellation was happening. but that would need to be done broad as the cancelllation can also happening on the notification step itself
@msujew any ideas?

**Reply by cdietrich** [reply]
potential workaround (helps in 98% of the cases)

```
  protected override async runCancelable(
    documents: LangiumDocument[],
    targetState: DocumentState,
    cancelToken: CancellationToken,
    callback: (document: LangiumDocument) => Promise<unknown>
  ): Promise<void> {
    const filtered = documents.filter(e => e.state < targetState)
    for (const document of filtered) {
      await interruptAndCheck(cancelToken)
      await callback(document)
    }
    await this.notifyBuildPhase(filtered, targetState, cancelToken)
    for (const document of filtered) {
      document.state = targetState
    }
    this.currentState = targetState
  }

  protected override async notifyBuildPhase(
    documents: LangiumDocument[],
    state: DocumentState,
    _cancelToken: CancellationToken
  ): Promise<void> {
    if (documents.length === 0) {
      // Don't notify when no document has been processed
      return
    }
    const listeners = this.buildPhaseListeners.get(state)
    for (const listener of listeners) {
      // await interruptAndCheck(cancelToken);
      await listener(documents, CancellationToken.None)
    }
  }
```
(just doing the state change at the end does not seem to be sufficient in all cases, i assume this is cause the references are resolved and thus no revalidation will be happening anyway)

**Reply by msujew** [reply]
I'm not sure this is the right approach:

1. This is way to broad, as that will also affect phases like parsing, and indexing, which should not be repeated for the same file. In fact, this mechanism would require us to "reset" the AST/index for a given file in case the build was interrupted on a different file for the same phase.
2. I'm still not sure why this actually fixes the issue. Invalid files (i.e. those with missing references) are supposed to be relinked anyway due to the `isAffected` computation. I'm not sure how preventing to advance the state in case of cancellation improves on this.

**Reply by cdietrich** [reply]
The problem is that the notification (that there are no markers anymore) won’t arrive . If cancel is in between 

I can provide the Java program to ease reproduction if it helps

**Reply by cdietrich** [reply]
```
package fuse_simulator;

import java.nio.charset.StandardCharsets;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.StandardOpenOption;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.google.common.collect.Lists;

public class FuseSimulator {
	
	public static void main(String[] args) throws Exception {
		String dir = "/Users/dietrich/Downloads/all_end_june/fuse";
		List<String> fileList = new ArrayList<>();
	    Files.walkFileTree(Paths.get(dir), new SimpleFileVisitor<Path>() {
	        @Override
	        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
	            if (!Files.isDirectory(file)) {
	                fileList.add(file.toString());
	            }
	            return FileVisitResult.CONTINUE;
	        }
	    });
	    Collections.shuffle(fileList);
	    List<List<String>> partitions = Lists.partition(fileList, 50);
	    for (List<String> part : partitions) {
	    	System.out.println(part);
	    	for (String file : part) {
	    		Path oldP = Paths.get(file);
	    		Path newP;
	    		if (file.endsWith("_bak")) {
	    			newP = Paths.get(file.substring(0,file.length()-4));
	    		} else {
	    			newP = Paths.get(file+"_bak");
	    		}
	    		List<String> allLines = Files.readAllLines(oldP);
	    		Files.delete(oldP);
	    		Files.createFile(newP);
	    		boolean ls = newP.toString().endsWith("_bak");
				List<List<String>> halves = Lists.partition(allLines, Math.max(1, (int)(allLines.size()/4)));
				for (List<String> linePart : halves) {
					for (String line : linePart) {
		    			Files.write(newP, (line+"\n").getBytes(StandardCharsets.UTF_8), StandardOpenOption.APPEND);
		    		}
					if (!ls && Math.random()<0.05) {
						Thread.sleep((int)(Math.random()*250));
					}
				}
	    		
	    	}
	    }
	}

}

```

this basically renames all `*.extension` to `*.extension_bak` and back to `*.extension`
assuming in the workspace there is only `.extension` files

**Reply by msujew** [reply]
> The problem is that the notification (that there are no markers anymore) won’t arrive . If cancel is in between

Ooooh, now I see. Yeah, that makes perfect sense. I know what to do now, thanks a lot for investigating this 👍
